Chore/some-fc

jackkav opened the pull request:
motivation; remove another 10% of autobind classes, learn and highlight examples for discussion of migration techniques for the following:
- ref forwarding to child component
- ref interaction
- set timeout

scope: 12 functional components tested and with limited impact, pagelayout takes render props rather than render functions, which allows us to nest FCs inside it without classes.

takeaways:
- createRef in the class and forwardRef alone in the FC is good when we don't need the forwarder to care about or override the ref functions.
- set Timeout can be replaced with useInterval
- `onClick(e=>fn(e))` and `onClick((e,v)=>fn(v))` and value={v} are used inconsistently on buttons and dropdown and we could find a more consistent approach.
- useCallback and function memoization still feels like premature optimisation.
- modals and code editor ref passing is still hard
- these changes should improve HMR dx
- our Button component extends HTMLButtonElement to add a function we don't use.
- our Button and Dropdown components are duplicated in insomnia-components and the ui/.../base folder, holding up any effective migration or improvements.
- we redefine FC prop types a lot, would be great to have more inference.
<!--
Please open an [Issue](https://github.com/kong/insomnia/issues/new) first to discuss new
features or non-trivial changes. Please provide as much detail as possible on the change as
possible including general description, implementation details, potential shortcomings, etc.

If this PR closes an issue, please mention "Closes #XX" where #XX is the issue number.

If this PR fixes a bug or regression, please make sure to add a test.
-->

jackkav committed:
editable

jackkav committed:
wrapper debug

jackkav committed:
method dropdown

jackkav committed:
request actions dropdown

jackkav committed:
response history dropdown

jackkav committed:
wrapper home

jackkav committed:
copy button

jackkav committed:
auth dropdown

jackkav committed:
env dropdown

jackkav committed:
content type dropdown

jackkav committed:
fix types

jackkav commented on packages/insomnia/src/ui/components/base/copy-button.tsx:
```diff
@@ -1,76 +1,54 @@
-          <span>
-            {confirm} <i className="fa fa-check-circle-o" />
-          </span>
-        ) : (
-          children || 'Copy to Clipboard'
-        )}
-      </Button>
-    );
-  }
-}
+    return () => clearTimeout(timer);
+  }, [showConfirmation]);
```
setTimeout example

jackkav commented on packages/insomnia/src/ui/components/dropdowns/method-dropdown.tsx:
```diff
@@ -18,16 +17,25 @@ interface Props {
+  showGrpc,
+  className,
+}, ref) => {
+  const dropdownRef = useRef<Dropdown>(null);
+  const toggle = useCallback(() => {
+    if (dropdownRef.current) {
+      dropdownRef.current.toggle();
+    }
+  }, [dropdownRef]);
+  useImperativeHandle(ref, () => ({ toggle }), [toggle]
+  );
```
feels like boilerplate and no code would be better if its possible, since we aren't doing anything special with it here except redefining toggle.

dimitropoulos replied:
can you make a commit (and maybe just push it to a temporary branch) to demonstrate what you're purposed solution is.  I'm not seeing how to make this better, boilerplate or not.

jackkav replied:
I haven't found anything that works yet, just an intuition.

dimitropoulos replied:
I share your sentiment, however as far as I'm aware from reading the docs this is "the way to do it:tm:"

marckong replied:
The doc also says, it should be avoided in most cases.

dimitropoulos replied:
yes, and what I was saying is that until someone provides another way to do it this is the case where it can't be avoided.  agreed.  seems like we found that "other way" in this case because the ref happens to already have had the extra stuff attached, but if we were to, for example, convert Dropdown to an FC, then we'd definitely have to use useImperativeHandle there in that component to keep the same or similar functionality we have now.

jackkav committed:
second pass

jackkav committed:
response csv viewer

jackkav committed:
useCallback pass

jackkav committed:
more useCallbacks

jackkav committed:
wrapper unit test

jackkav committed:
simplify pagelayout

filfreire reviewed (approved):
LGTM, there are some bugs though ⚠️ 


- 🐞 When you double click now to edit a test name, and you write down a new name, after it goes out of focus the name disappears:
<img width="349" alt="image" src="https://user-images.githubusercontent.com/11976836/176146461-1ce899d7-3da2-43a6-8ec4-8026d8ae7581.png">

(only seems to happen with the editable on test names, other editables seem to work)


Also tested the copy button, and the different dropdown and they seem to be ok. Csv viewer also seems to be ok.

jackkav committed:
fix editable bug in unit test

jackkav committed:
wrapper unit test useCallbacks

jackkav committed:
some gentle tidying

marckong commented on packages/insomnia/src/ui/components/request-url-bar.tsx:
```diff
@@ -58,7 +58,7 @@ export class RequestUrlBar extends PureComponent<Props, State> {
   }
 
-  _setMethodDropdownRef(methodDropdown: MethodDropdown) {
+  _setMethodDropdownRef(methodDropdown: MethodDropdownHandle) {
```
have we tried using createRef() and forward it to the dropdown and use dropdown directly instead of using it with useImperativeHandle?

marckong commented on packages/insomnia/src/ui/components/request-url-bar.tsx:
```diff
@@ -45,7 +45,7 @@ export class RequestUrlBar extends PureComponent<Props, State> {
   _sendInterval: NodeJS.Timeout | null = null;
   _dropdown: Dropdown | null = null;
-  _methodDropdown: MethodDropdown | null = null;
+  _methodDropdown: MethodDropdownHandle | null = null;
```
I wonder if we can do it like the following instead:
```js
_dropdownRef = createRef<Dropdown>();
```

and then,
```js
<MethodDropdown ref={this._dropdownRef} />
```

marckong commented on packages/insomnia/src/ui/components/dropdowns/method-dropdown.tsx:
```diff
@@ -18,16 +17,25 @@ interface Props {
+  onChange,
+  showGrpc,
+  className,
+}, ref) => {
```
if we create a ref from where this is used as Dropdown ref, you can possibly forward it directly and use it without useImperativeHandle.

dimitropoulos committed:
useCallback usage for async func

dimitropoulos committed:
switch to useInterval

dimitropoulos committed:
updates types for Editable, Highlight, et al.

dimitropoulos committed:
moves usage after declaration

dimitropoulos committed:
extracts FC for AuthDropdown items

dimitropoulos committed:
removes some prop drilling and extracts component

dimitropoulos committed:
adds mandaotry useCallback, rm unused props, types

dimitropoulos committed:
use named exports from react npm like rest of project

eventually, we will remove usages of `React` default export since it's a hack from half a decade ago that's no longer necessary now that jsx-transform is a thing.

dimitropoulos committed:
removes unnecessary useImperativeHandle (we learn somethin' every day)

dimitropoulos committed:
adds necessary useCallback, removes unnecessary useImparativeHandle, TS

dimitropoulos committed:
fixes types for useRef (turns out you don't need null in the generic)

dimitropoulos committed:
removes unused props

dimitropoulos committed:
fixes types from ref helpers

dimitropoulos committed:
fixes return from useCallback

dimitropoulos committed:
updates useCallback usages

dimitropoulos committed:
fixes bind

dimitropoulos committed:
fixes lint

for some reason, kept getting:

```
insomnia: /home/dimitri/src/github.com/Kong/insomnia/packages/insomnia/src/preload.js
insomnia:   0:0  error  Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
insomnia: The file does not match your project config: packages/insomnia/src/preload.js.
insomnia: The file must be included in at least one of the projects provided
```

dimitropoulos commented:
I'm about an hour later working than I was aiming for so I had to stop but I did a thorough review and pushed updates for what I found - lemme know your thoughts :)

I would have loved to test it more (I did a little as I went, but not much since I was planning to do so at the end.... but then I ran out of time).

jackkav committed:
revert incomplete type change

jackkav commented on packages/insomnia/src/ui/components/base/copy-button.tsx:
```diff
@@ -16,17 +16,20 @@ export const CopyButton: React.FC<Props> = ({
-    event.preventDefault();
-    event.stopPropagation();
-    const toCopy = typeof content === 'string' ? content : await content();
+  const onClick = useCallback((event: React.MouseEvent) => {
```
non-blocking: its not clear to me why we shouldn't use `useCallback(async(event: React.MouseEvent)` here

jackkav commented on packages/insomnia/src/ui/components/wrapper-home.tsx:
```diff
@@ -159,13 +157,33 @@ const WrapperHome: FC<Props> = (({ wrapperProps }) => {
-  const handleImportClipboard = useCallback(({ forceToWorkspace }) => dispatch(importClipBoard({ forceToWorkspace })), [dispatch]);
-  const handleSetDashboardSortOrder = useCallback(sortOrder => dispatch(setDashboardSortOrder(sortOrder)), [dispatch]);
-  const handleActivateWorkspace = useCallback(({ workspace }) => dispatch(activateWorkspace({ workspace })), [dispatch]);
+  const handleCreateWorkspace = useCallback(({ scope, onCreate }) => {
+    dispatch(createWorkspace({ scope, onCreate }));
```
non-blocking: I'd like to understand why returning the dispatch is no bueno

jackkav commented on packages/insomnia/src/ui/components/wrapper-unit-test.tsx:
```diff
@@ -209,8 +212,8 @@ const WrapperUnitTest: FC<Props> = ({
-    await models.unitTest.update(unitTest, { code });
+  const handleUnitTestCodeChange = useCallback((unitTest: UnitTest, code: string) => {
+    models.unitTest.update(unitTest, { code });
   }, []);
```
very cool

jackkav commented on packages/insomnia/src/ui/components/dropdowns/request-actions-dropdown.tsx:
```diff
@@ -26,10 +26,10 @@ import { showError } from '../modals';
-  handleDuplicateRequest: (reqeust: Request | GrpcRequest) => void;
-  handleGenerateCode: (reqeust: Request | GrpcRequest) => void;
-  handleSetRequestPinned: (reqeust: Request | GrpcRequest, isPinned: boolean) => void;
+  handleSetRequestPinned: Function;
+  handleDuplicateRequest: Function;
+  handleGenerateCode: Function;
+  handleCopyAsCurl: Function;
```
I tried to fix the type issues with this but it scope creeped into logical changes because `handleActivateRequest: (requestId?: string) => void`
is typed as `handleActivateRequest: (activeRequestId: string) => void` in wrapper props where is comes from and I didn't want the types to influence behaviour in this PR scope. Cool to do in a follow up though. The more type inference the better!

gatzjames commented on packages/insomnia/src/ui/components/modals/analytics-modal.tsx:
```diff
@@ -74,7 +74,7 @@ const ActionButtons = styled.div({
 export const AnalyticsModal: FC = () => {
   const { hasPromptedAnalytics } = useSelector(selectSettings);
-  const ref = useRef<Modal | null>(null);
+  const ref = useRef<Modal>(null);
```
👌

gatzjames commented on packages/insomnia/src/ui/components/page-layout.tsx:
```diff
@@ -20,11 +20,11 @@ const Pane = forwardRef<HTMLElement, { position: string; children: ReactNode }>(
-  renderPageBody?: () => ReactNode;
-  renderPaneOne?: () => ReactNode;
-  renderPaneTwo?: () => ReactNode;
+  renderPageSidebar?: ReactNode;
```
Since these aren't render functions we could rename them to pageSidebar, pageHeader etc.

gatzjames commented on packages/insomnia/src/ui/components/page-layout.tsx:
```diff
@@ -68,13 +65,12 @@ export const PageLayout: FC<Props> = ({
   const realSidebarWidth = sidebarHidden ? 0 : sidebarWidth;
-  const paneTwo = renderPaneTwo?.();
-  const gridRows = paneTwo
+  const gridRows = renderPaneTwo
```
🙏

gatzjames commented on packages/insomnia/src/ui/components/viewers/response-csv-viewer.tsx:
```diff
@@ -1,73 +1,40 @@
-    Papa.parse<string[]>(csv, {
+export const ResponseCSVViewer: FC<Props> = ({ body }) => {
+  const [csv, setCSV] = useState<{ data: string[][] } | null>(null);
+  const parse = useCallback(() => {
```
This could be changed to useEffect since it's not used in an event handler.

```suggestion
const parse = useEffect(() => {
```

gatzjames reviewed (approved):
Looks nice! 🚀👍

Tested and left some comments but nothing blocking!

jackkav committed:
remove unnecessary useEffect

gatzjames commented on packages/insomnia/tsconfig.json:
```diff
@@ -10,18 +10,20 @@
     "esbuild.sr.ts",
     "jest.config.js",
     "package.json",
+    "preload.js",
```
🤔 This looks a bit confusing since we include preload.js and exclude src/preload.js
